src/lib/remark-plugins/rewrite-tutorial-links/index.ts

```diff
@@ -114,45 +129,63 @@ export function rewriteTutorialsLink(
 		}
 
 		if (isBetaProduct || isValidSection) {
-			let nodePath = url // the path to be formatted - assumes to be absolute as current Learn impl does
-			const isCollectionPath = nodePath.includes('collections')
-			const isTutorialPath = nodePath.includes('tutorials')
-			const learnProductHub = new RegExp(`/${product}$`)
-			const isProductHubPath = learnProductHub.test(nodePath)
-			const isDocsPath = nodePath.includes(productSlugsToHostNames[product])
-
-			// if its an external link, isolate the pathname
-			if (isExternalLearnLink || isDocsPath) {
-				const fullUrl = new URL(nodePath)
-				// removing the origin from the href instead of only using
-				// 'pathname' so that anchor links are included
-				nodePath = fullUrl.href.replace(fullUrl.origin, '')
+			const urlObject = new URL(url, 'https://learn.hashicorp.com/')
+			const { origin, pathname } = urlObject
+			const urlWithoutOrigin = urlObject.toString().replace(origin, '')
+
+			const collectionPathRegex = new RegExp('^/collections')
+			const isCollectionPath = collectionPathRegex.test(pathname)
+			if (isCollectionPath) {
+				newUrl = handleCollectionLink(urlWithoutOrigin)
+			}
+
+			const tutorialPathRegex = new RegExp('^/tutorials')
+			const isTutorialPath = tutorialPathRegex.test(pathname)
+			if (isTutorialPath) {
+				newUrl = handleTutorialLink(urlWithoutOrigin, tutorialMap)
 			}
 
-			// handle rewriting collection and tutorial dev portal paths
+			const productHubPathRegex = new RegExp(`^/${product}/?$`)
+			const isProductHubPath = productHubPathRegex.test(pathname)
+			if (isProductHubPath) {
+				newUrl = `/${product}/tutorials`
+			}
+
+			const productIOHostName = productSlugsToHostNames[product]
+			const isDocsPath = origin.includes(productIOHostName)
 			if (isDocsPath) {
-				newUrl = handleDocsLink(nodePath, product as ProductSlug)
-			} else if (isCollectionPath) {
-				newUrl = handleCollectionLink(nodePath)
-			} else if (isTutorialPath) {
-				newUrl = handleTutorialLink(nodePath, tutorialMap)
-			} else if (isProductHubPath) {
-				newUrl = `${nodePath}/tutorials`
+				newUrl = handleDocsLink(urlWithoutOrigin, product as ProductSlug)
+			}
+
+			/**
+			 * Check if multiple conditions above were true. Only one should be true
+			 * at a time.
+			 */
+			const isUrlAmbiguous = getIsUrlAmbiguous({
+				isCollectionPath,
+				isTutorialPath,
+				isProductHubPath,
+				isDocsPath,
+			})
+			if (isUrlAmbiguous) {
+				throw new Error(`[rewriteTutorialsLink] found an ambiguous url: ${url}`)
```

Comment:
If it is ambiguous, do we want to adjust what is returned, for example we may not want to rewrite it at all? Or do we allow for it to be rewritten from whichever condition last caught it?

Reply:
I updated the function to return `newUrl || url`, so the original `url` is what should be returned in this case.

Reply:
🤔 well if the `newUrl` was set in one of those conditionals, then `url` wouldn't return right?

Reply:
Ah right, since this check is happening after the conditionals. Let me do a reorder of things real quick so that doesn't happen.

Reply:
It could be good to add a test case for an ambiguous url perhaps?

Reply:
Just pushed the re-ordering update in https://github.com/hashicorp/dev-portal/pull/1014/commits/ae6f214b84ab8ac7487aa9f372545ad5145b2c5b. I'm updating docs sidebar to handle external Learn links tomorrow, and I'm going to add more tests and refactor a few things here as I work on it. It will make it a lot easier to come up with the ambiguous url test case. With that fast-follow in mind, are you okay with merging this without the ambiguous url test case?